# Employee Review Demo

A demo of using [SurrealDB](https://surrealdb.com/) to query Kaggle's [Employee Review](https://www.kaggle.com/datasets/fiodarryzhykau/employee-review) dataset.

In [80]:
%pip -q install surrealdb pandas

Note: you may need to restart the kernel to use updated packages.


## Get started

### Run database in container

First, run SurrealDB as a container. I use [podman](https://podman.io/) because I like it better, but docker will work too.

The volume mount is optional, in case you want to export your database.

```{danger}
Passing credentials like this is insecure and is for testing and demo only!
```

```bash
podman run --rm -it -p 8000:8000 -v `pwd`/mydata:/mydata docker.io/surrealdb/surrealdb:latest start --auth --user root --pass testing 
```

You should see

> INFO surrealdb::net: Started web server on 0.0.0.0:8000

### Login to database

We will use the [Python SDK for SurrealDB](https://docs.surrealdb.com/docs/integration/sdks/python/).

Note that I believe Python is an awful language and you should avoid it in most production cases. But for this demo notebook, eh, I suppose I can hold my nose and do it.

SurrealDB has a plethora of SDKs to chose from, though they aren't as developed as one might hope.

In [81]:
from surrealdb import Surreal

db = Surreal("http://localhost:8000")
await db.connect()

# NOTE: This is an insecure way of handling credentials and should not be used in production.
await db.signin({"user": "root", "pass": "testing"})
await db.use("test", "test")

print(db.client_state)

ConnectionState.CONNECTED


## Preview data

Again, this dataset is from Kaggle's [Employee Review](https://www.kaggle.com/datasets/fiodarryzhykau/employee-review).

Let's check it out. We'll also grab a list of unique names because we will need those.

In [82]:
# Preview the CSV
import pandas as pd

# Load the CSV file
file_path = 'augmented_employee_feedback.csv'
data = pd.read_csv(file_path)

# Display the info about the dataframe
print(data.info())

# For each unique employeeId in the CSV
unique_employeeId = data['employeeId'].unique()
print("Unique employees in dataset:", len(unique_employeeId))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 225 entries, 0 to 224
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   employeeId  225 non-null    int64 
 1   name        225 non-null    object
 2   year        225 non-null    int64 
 3   team        225 non-null    object
 4   nine_box    225 non-null    int64 
 5   feedback    225 non-null    object
 6   Source      225 non-null    object
dtypes: int64(3), object(4)
memory usage: 12.4+ KB
None
Unique employees in dataset: 75


## Create member table

The first thing we will do with our database is create table to put the members of our company in. Let's call it **member**.

We will use a [schemafull table](https://docs.surrealdb.com/docs/surrealql/statements/define/table#schemafull-tables)
because then the database will strictly enforce the definitions we add to the table for each record.

While schemaless tables offer advantages in flexibility, for something as critical as the cornerstone record of who an employee is, schemafull is *absolutely* the way to go.

The queries can get a tad unwieldy, so I like to preview the string before sending to the database.

In [83]:
member_tb = "member"
createMemberQuery = f"""
# Create primary table for holding member info
DEFINE TABLE {member_tb} SCHEMAFULL;

# Name
DEFINE FIELD name ON TABLE {member_tb} TYPE string;

# Employee ID as unique index
DEFINE FIELD employeeId ON TABLE {member_tb} TYPE int
    ASSERT employeeId > 10000 && employeeId < 99999;
DEFINE INDEX employeeIdIndex ON TABLE {member_tb} COLUMNS employeeId UNIQUE;
"""

print(createMemberQuery)


# Create primary table for holding member info
DEFINE TABLE member SCHEMAFULL;

# Name
DEFINE FIELD name ON TABLE member TYPE string;

# Employee ID as unique index
DEFINE FIELD employeeId ON TABLE member TYPE int
    ASSERT employeeId > 10000 && employeeId < 99999;
DEFINE INDEX employeeIdIndex ON TABLE member COLUMNS employeeId UNIQUE;



In [84]:
# If the command to create the member table looks good, go ahead and create it.
await db.query(createMemberQuery)
# Verify the table was created
await db.query("INFO FOR TABLE member;")

[{'result': {'events': {},
   'fields': {'employeeId': 'DEFINE FIELD employeeId ON member TYPE int ASSERT employeeId > 10000 AND employeeId < 99999 PERMISSIONS FULL',
    'name': 'DEFINE FIELD name ON member TYPE string PERMISSIONS FULL'},
   'indexes': {'employeeIdIndex': 'DEFINE INDEX employeeIdIndex ON member FIELDS employeeId UNIQUE'},
   'lives': {},
   'tables': {}},
  'status': 'OK',
  'time': '2.672508ms'}]

### employeeId

Let's randomly assign each employee an ID number. Recall that the schema definition above enforces that `employeeId` is

- **UNIQUE** amongst all `employeeIds`
- A string made up of 5 digits
- [Indexed](https://docs.surrealdb.com/docs/surrealql/statements/define/indexes) with `employeeIdIndex`.

The record creation here will fail if any of those criteria (as well as the other definitions) are not met.

In [85]:
from random import randint

# Roster of entries
records_added = 0
failed_records: list[dict[str, str]] = []

# Iterate through the unique employee IDs in data and get the name
for employee_id in unique_employeeId:
    # Get the name for the employee ID
    name = data.loc[data['employeeId'] == employee_id, 'name'].iloc[0]
    print(name, employee_id)
    # Add the person to the member table
    try:
        result = await db.create(member_tb, {
        'name': name,
        'employeeId': int(employee_id),
        })
        # If result contains 'status': 'ERR' then throw an exception
        if result[0]["status"] == "ERR":
            raise Exception(result)
        records_added = records_added + 1
    except Exception as e:
        failed_records.append({name, e})
        break

print("Added records:", records_added)
print("Errors adding:", len(failed_records))
print("Failed records:\n", failed_records)

Lacey Howard 63379
Amy Jones 17072
Rylan Mack 18337
Blake Booth 73822
Brayden Harding 47392
Bryan Murray 99053
Zoe White 37374
Valeria Crane 66919
Johanna Duke 47631
Jon Fischer 75658
Noah Poole 63413
Max Miller 74495
Georgia Rogers 97447
Milania Hodge 73234
Alisa Stark 31729
Ashton Owen 46156
Dominic Wood 26581
Niamh Walsh 40186
Ella Powers 17594
Regan Rhodes 26681
Talon Miller 55760
Ivan Reese 12605
Seamus Noble 61175
Ella Green 13965
Scott Phillips 96638
Logan Mason 36223
Xavier Mcdowell 95838
Daisy Pearce 17293
Deon Griffith 11287
Bruno Johns 39262
Conner Mcintyre 37401
Benjamin Roberts 66313
Freya Price 24064
Matthew Reid 78305
Angelica Peterson 88056
David Kelly 19007
Archie Dawson 55723
Rowan Cain 45455
Chelsea Jimenez 43941
Zachary Turner 93064
Aryanna Carney 47590
Libby Parker 73688
Ewan Pearson 28614
Rachel Harper 25771
Jack Walsh 60326
Patrick Adams 98891
Allan Logan 90861
Layne Terrell 24427
Stephen Scott 76053
Aydin Pitts 68514
Andrew Grant 58424
Michael Kaur 74685
Dylan B

In [86]:
# View some of the results
await db.query(f"SELECT * FROM {member_tb} LIMIT 5;")

[{'result': [{'employeeId': 25771,
    'id': 'member:03048b1kicsk3p6hkfrs',
    'name': 'Rachel Harper'},
   {'employeeId': 53702,
    'id': 'member:04gjfhr661zwmnv3808v',
    'name': 'Finley Gray'},
   {'employeeId': 96638,
    'id': 'member:06zzan3g7g0dv2c8mvkf',
    'name': 'Scott Phillips'},
   {'employeeId': 75658,
    'id': 'member:0f83v9ysdzae76dpfshd',
    'name': 'Jon Fischer'},
   {'employeeId': 95732,
    'id': 'member:17bl9bm1z1yqz55n97sk',
    'name': 'Kyle Shaw'}],
  'status': 'OK',
  'time': '78.12µs'}]

## Add performance feedback

Now that we have our employees loaded in to the database, let's add the performance feedback!

### Create feedback table

This table will be a schemafull **edge** table: **gotFeedback**. I advocate for schemafull here because I'm imagining a standard feedback form with fields that must be filled in.

Additionally, there are some advantages for queries if you standardize fields. We will see this in action below.

In [87]:
# Create a schemafull table for feedback
feedback_tb = "gotFeedback"
createFeedbackQuery = f"""
DEFINE TABLE {feedback_tb} SCHEMAFULL;
DEFINE FIELD date on TABLE {feedback_tb} TYPE datetime;
DEFINE FIELD body ON TABLE {feedback_tb} TYPE string;
DEFINE FIELD rating ON TABLE {feedback_tb} TYPE number;
DEFINE FIELD team ON TABLE {feedback_tb} TYPE record<team>;
"""

print(createFeedbackQuery)


DEFINE TABLE gotFeedback SCHEMAFULL;
DEFINE FIELD date on TABLE gotFeedback TYPE datetime;
DEFINE FIELD body ON TABLE gotFeedback TYPE string;
DEFINE FIELD rating ON TABLE gotFeedback TYPE number;
DEFINE FIELD team ON TABLE gotFeedback TYPE record<team>;



In [88]:
# If the command to create the table looks good, go ahead and create it.
await db.query(createFeedbackQuery)
# Verify the table was created
print(await db.query(f"INFO FOR TABLE {feedback_tb};"))

[{'result': {'events': {}, 'fields': {'body': 'DEFINE FIELD body ON gotFeedback TYPE string PERMISSIONS FULL', 'date': 'DEFINE FIELD date ON gotFeedback TYPE datetime PERMISSIONS FULL', 'rating': 'DEFINE FIELD rating ON gotFeedback TYPE number PERMISSIONS FULL', 'team': 'DEFINE FIELD team ON gotFeedback TYPE record<team> PERMISSIONS FULL'}, 'indexes': {}, 'lives': {}, 'tables': {}}, 'status': 'OK', 'time': '129.248µs'}]


#### Define team table

We also happen to need a team table

In [89]:
# Define a schemaless team table
await db.query("DEFINE TABLE team SCHEMALESS;")

# Iterate over the teams and add them to the team table
for team in data['team'].unique():
    try:
        await db.create("team:" + team, {
            'name': team,
        })
    except Exception as e:
        print(e)
        break

### Add Feedback

To add the feedback to our database we'll make a few assumptions. Specifically, we'll randomly select and assign a manager and date.

In [90]:
# Assume managers are the last 5 names in the list
managers = unique_names[-5:]
print(managers)

['Zachary Doyle' 'Lauren Baker' 'George Jones' 'Aubri Hartman'
 'Ivanna Boyer']


In [95]:
from random import choice

# Go through each feedback and add it to the database with a randomly selected manager
feedback_df = data[['employeeId','feedback', 'team', 'year', 'nine_box']]
# Roster of entries
records_added = 0
failed_records: list[dict[str, str]] = []

# Add feedback
for _, row in feedback_df.iterrows():
    # Randomly generate datetime for when feedback happened since dataset only has year
    feedbackDate = str(row.year) + "-" + str(randint(10,12)) + "-" + str(randint(10,28))
    feedbackDate = feedbackDate + "T00:00:00Z"
    try:
        # Randomly select rater and get their member id
        rater = choice(managers)
        raterId = await db.query("SELECT id FROM member WHERE name = $rater;", {'rater':rater})
        raterId = raterId[0]["result"][0]["id"]
        # Get id of member
        memberId = await db.query(f"""SELECT id FROM member WHERE employeeId = {row.employeeId};""")
        memberId = memberId[0]["result"][0]["id"]
        query = f"""RELATE {memberId}->gotFeedback->{raterId} CONTENT {{
                    'date': "{feedbackDate}",
                    'rating': {row.nine_box},
                    'team': team:{row.team},
                    'body': "{row.feedback.replace('"', "'")}",   
                  }};
                  """
        result = await db.query(query)
        # If result contains 'status': 'ERR' then throw an exception
        if result[0]["status"] == "ERR":
            raise Exception(result)
        
        records_added = records_added +1
    except Exception as e:
        failed_records.append({name, e})

print("Added records:", records_added)
print("Errors adding:", len(failed_records))
print("Failed records:\n", failed_records)

    

Added records: 225
Errors adding: 0
Failed records:
 []


## Query Feedback Records

Get all feedback a person has ever gotten

- `SELECT ->gotFeedback.out FROM member:v8hze8yz5emd3occkuoq;` will iterate the member object: *fastest*
- `SELECT ->gotFeedback.out FROM member WHERE employeeId = 99053;` will iterate **employeeIdIndex**: fast

Get all feedback a person has ever given

- `SELECT <-gotFeedback.in FROM member:k24bvwin6pr3wlkrs4lm;` will iterate the member object: *fastest*
- `SELECT <-gotFeedback.in FROM member WHERE name = "Zachary Doyle";` will iterate the **member** table: painfully slow

---

### Query Feedback Records

In this example we'll compare a traditional table query with a graph query. The objective is to get the text and rater for each and every feedback an employee has ever received.

Remember that our database currently has:
- A `member` table that holds basic information about employees.
- A `feedback` table that holds a record for every feedback.
- A `gotFeedback` [edge table](https://docs.surrealdb.com/docs/surrealql/statements/relate) that points from an employee to their feedback.

Furthermore, our `member` table has an INDEX on the `employeeId` column.

We will use **Valeria Crane** as our example employee. First, let's get her employee ID.
This would be known in a production environment - probably even printed on the back of a badge - so
it makes sense to somewhat artificially get it here.

In [186]:
# Get information about employee named Valeria Crane
crane = await db.query("SELECT * FROM member WHERE name = 'Valeria Crane'")
crane_employeeId = crane[0]['result'][0]['employeeId']
print("Crane's employeeId:", crane_employeeId)

Crane's employeeId: 65810


Run the table query first. This matches the approach taken by a traditional relational database.
Notice the `time` output at the bottom (about 3 milliseconds on my computer).

In [187]:
# Give me the rater name and the body of the feedback for every feedback the employee has gotten
crane_feedback_table = f"""SELECT rater.name AS rater, body FROM feedback WHERE ratee.employeeId = '{crane_employeeId}'"""

await db.query(crane_feedback_table)

[{'result': [{'body': 'Valeria Crane is a hazard in our team. She has shown no signs of reliability. She is difficult to work with and does not seem motivated. Valeria needs a lot of guidance to complete tasks.',
    'rater': 'Heidi Wallace'},
   {'body': 'To simply but, Valeria is currently ranked last in the customer service leaderboard, having missed some basic but important sales target and slacking off at work. She is constantly on her phone to the point where we had to bar the use of phones within the office, causing inconvenience to genuine team members. I am afraid, unless she improves, she may be counting her last few weeks in this company.',
    'rater': 'Jaxson Giles'},
   {'body': 'Valeria is considered a risk. She has not been performing up to required standards. We have tried working with her to improve and there has not been any improvement. More testing will be required, but I would not recommend for personal advancement until noticeable and consistent improvement from 

Now get the same information with a graph query. The output format will look slightly different, but it's still JSON.
Check out the `time` though! On my computer it's about 500 microseconds, or **6x as fast!**

In [188]:
# Get all of Crane's feedback using a graph query
crane_feedback_graph = f"""SELECT ->gotFeedback.out.rater.name AS rater, ->gotFeedback.out.body AS body
                            FROM (SELECT id FROM member WHERE employeeId = '{crane_employeeId}')"""
await db.query(crane_feedback_graph)


[{'result': [{'body': ['Valeria Crane is a hazard in our team. She has shown no signs of reliability. She is difficult to work with and does not seem motivated. Valeria needs a lot of guidance to complete tasks.',
     'Valeria Crane has shown little in her work to amaze. So far it has been of a level below standard. We have not seen a consistent member of the team. Undertaking tasks has been a poor turn out so far from Valeria.',
     'To simply but, Valeria is currently ranked last in the customer service leaderboard, having missed some basic but important sales target and slacking off at work. She is constantly on her phone to the point where we had to bar the use of phones within the office, causing inconvenience to genuine team members. I am afraid, unless she improves, she may be counting her last few weeks in this company.',
     'Valeria is considered a risk. She has not been performing up to required standards. We have tried working with her to improve and there has not been a

### Explain the difference

So what's going on here? None of these tables are huge (member=75 rows, feedback=225 rows) so I was surprised to see these dramatic of results.

Well, let's use the [EXPLAIN](https://docs.surrealdb.com/docs/surrealql/statements/select/#the-explain-clause) clause.

#### Table Query

You'll see the the table query requires the database to **Iterate Table** over the *entire* **feedback** table because **NO INDEX FOUND**.

Now, we *could* add another column into the feedback table that includes `employeeId` and then make that an index. However, this means that INSERT would be slower because you have to update the index.

Additionally, as we will see, you don't actually need it; you can use a graph query instead!

In [189]:
from json import dumps # formats printing of json

print((crane_feedback_table))
# Explain how the query was preformed
print("Explanation for table lookup by name:")
print(dumps(await db.query(crane_feedback_table + " EXPLAIN"), indent=4))

SELECT rater.name AS rater, body FROM feedback WHERE ratee.employeeId = '65810'
Explanation for table lookup by name:
[
    {
        "result": [
            {
                "detail": {
                    "table": "feedback"
                },
                "operation": "Iterate Table"
            },
            {
                "detail": {
                    "reason": "NO INDEX FOUND"
                },
                "operation": "Fallback"
            }
        ],
        "status": "OK",
        "time": "41.105\u00b5s"
    }
]


#### Graph Query

First, notice that the graph query actually requires two queries.

1. Get the `member` object that has the matching `employeeId` (this is the rightmost query, in parenthesis).
2. Get the information from the **feedback** table.

In a production environment the first query is unlikely to be necessary because you probably already have the member ID from authentication or some other request. But leaving it here is educational.

The first EXPLAIN states that the database uses **Iterate Index** on the **employeeIdIndex** from the **member** table to quickly get the `member` object. This is a good-old-fashioned relational query with an index. It's wicked fast and showcases exactly why we have indexes in the first place.

The second query is the graph query (notice the `->` symbols). That operation is **Iterate Value** on the `member` object.

So, what does this mean?

The query works by tracing edges held in the **gotFeedback** table. By tracing the edges to the appropriate records in the **feedback** table, we neither had to iterate over the table nor build an index to make that iteration faster.

The consequence is that unless an employee has a huge number of feedback entries, it will be much faster to just use our edge table. Even better, we won't experience overhead rebuilding an index on an INSERT to **feedback**.

In [190]:
print(crane_feedback_graph)

print("Explanation for graph lookup:")
# Need two explain statements because there are actually two queries happening
print(dumps(await db.query(crane_feedback_graph.replace(')', ' EXPLAIN) EXPLAIN')), indent=4))

SELECT ->gotFeedback.out.rater.name AS rater, ->gotFeedback.out.body AS body
                            FROM (SELECT id FROM member WHERE employeeId = '65810')
Explanation for graph lookup:
[
    {
        "result": [
            {
                "detail": {
                    "value": {
                        "detail": {
                            "plan": {
                                "index": "employeeIdIndex",
                                "operator": "=",
                                "value": "65810"
                            },
                            "table": "member"
                        },
                        "operation": "Iterate Index"
                    }
                },
                "operation": "Iterate Value"
            }
        ],
        "status": "OK",
        "time": "57.603\u00b5s"
    }
]
